controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,466 @@
+	// changes to LastUpdated will not increase the resource generation
+	if e.ObjectNew.GetGeneration() != e.ObjectOld.GetGeneration() {
+		r.Log.Info("hostfirmwaresettings resource generation changed, updating")
+		return true
+	}
+
+	return false
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *HostFirmwareSettingsReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&metal3v1alpha1.HostFirmwareSettings{}).
+		WithEventFilter(
+			predicate.Funcs{
+				UpdateFunc: r.updateEventHandler,
+			}).
+		Watches(&source.Kind{Type: &metal3v1alpha1.BareMetalHost{}}, &handler.EnqueueRequestForObject{}, builder.Predicates{}).
+		Complete(r)
+}
+
+// Validate the HostFirmwareSetting Spec against the schema
+func (r *HostFirmwareSettingsReconciler) validateHostFirmwareSettings(info *rInfo, schema *metal3v1alpha1.FirmwareSchema) []error {
+
+	var errors []error
+
+	for name, val := range info.hfs.Spec.Settings {
+		// Prohibit any Spec settings with "Password"
+		if strings.Contains(name, "Password") {
+			errors = append(errors, fmt.Errorf("Cannot set Password field"))
+			break
+		}
+
+		// The setting must be in the Status
+		if _, ok := info.hfs.Status.Settings[name]; !ok {
+			errors = append(errors, fmt.Errorf("Setting %s is not in the Status field", name))
+			break
+		}
+
+		// check validity of updated value
+		if schema != nil {
+			if err := schema.ValidateSetting(name, val, schema.Spec.Schema); err != nil {
+				errors = append(errors, err)
+			}
+		}
+	}
+
+	if len(errors) > 0 {
+		return errors
+	}
+
+	return nil
+}
+
+func (r *HostFirmwareSettingsReconciler) publishEvent(request ctrl.Request, event corev1.Event) {
+	reqLogger := r.Log.WithValues("hostfirmwaresettings", request.NamespacedName)
+	reqLogger.Info("publishing event", "reason", event.Reason, "message", event.Message)
+	err := r.Create(context.TODO(), &event)
+	if err != nil {
+		reqLogger.Info("failed to record event, ignoring",
+			"reason", event.Reason, "message", event.Message, "error", err)
+	}
+	return
+}
+
+func setCondition(generation int64, status *metal3v1alpha1.HostFirmwareSettingsStatus,
+	cond metal3v1alpha1.SettingsConditionType, newStatus metav1.ConditionStatus,
+	reason conditionReason, message string) {
+	newCondition := metav1.Condition{
+		Type:               string(cond),
+		Status:             newStatus,
+		ObservedGeneration: generation,
+		Reason:             string(reason),
+		Message:            message,
+	}
+	meta.SetStatusCondition(&status.Conditions, newCondition)
+}
+
+// Generate a name based on the schema keys which should be the same for similar hardware
+func GetSchemaName(schema map[string]metal3v1alpha1.SettingSchema) string {
+
+	// Schemas from the same vendor and model should be identical for both keys and values.
+	// Hash the keys of the map to get a identifier unique to this schema.
```

Comment:
Why not include the values as well? Just having the keys sounds like it would risk collisions between different firmware versions.

Reply:
I'm not sure if we would practically see differences in values but I'll add the values into the hash to prevent that happening.